src/HDF5.jl

```diff
@@ -1395,7 +1395,13 @@ function read{S<:String}(obj::DatasetOrAttribute, ::Type{Array{S}})
             end
         end
     end
-    h5t_close(memtype_id)
+    try
+        dspace = dataspace(obj)
```

Comment:
I don't know if we care about performance, but if we do: When we call `size(obj)` we're also constructing a dataspace object, and we don't need to do it twice.

Reply:
I don't see an obvious way of returning the dataspace object that  `size` uses  so that It could be used instead of creating another `dataspace` object here.